src/Servers/Kestrel/Core/src/Internal/Infrastructure/Heartbeat.cs

```diff
@@ -66,7 +68,7 @@ internal void OnHeartbeat()
             {
                 if (!_debugger.IsAttached)
                 {
-                    _trace.HeartbeatSlow(_interval, now);
+                    _trace.HeartbeatSlow(now - _lastHeartbeat, _interval, now);
```

Comment:
I know `_lastHeartbeat` is only used for logging, but unsynchronized access could lead to tearing on 32-bit machines if another thread is in the middle of assigning `_lastHeartbeat` to `now`.

You can avoid this by storing `_lastHeartbeat` as an long and using `Volatile.Read/Write` like we do in the [HeartbeatManager](https://github.com/aspnet/AspNetCore/blob/master/src/Servers/Kestrel/Core/src/Internal/Infrastructure/HeartbeatManager.cs#L31).

Reply:
Can you just mark DateTimeOffset as volatile?

Reply:
I don't think the compiler will let you mark a DateTimeOffset as volatile since it is a struct. You could mark the ticks stored as a long as volatile though, and that should guarantee atomic reads and writes. That way you don't need to explicitly call `Volatile.Read/Write`.

> First off, so as to not bury the lead: in C# the rules have been carefully designed so that every volatile field read and write is also atomic. (Of course the converse does not follow; it is perfectly legal for an operation to be atomic without it being "volatile", whatever that means.)

https://blogs.msdn.microsoft.com/ericlippert/2011/06/16/atomicity-volatility-and-immutability-are-different-part-three/

Reply:
> You could mark the ticks stored as a long as volatile though, and that should guarantee atomic reads and writes.

Yep, this is generally the pattern used to store a date atomically. If you're trying to accurately store a `DateTimeOffset` though you need to handle storing the offset (which is a separate number). Or, you could just say it's OK for it to be UTC always and be fine with that.